File: Hexes/src/marcos/games/hexes/swing/HexesApplet.java
package marcos.games.hexes.swing;

import java.io.File;
import java.util.LinkedList;

import javax.swing.JOptionPane;

import cc.lib.game.AGraphics;
import cc.lib.game.GColor;
import cc.lib.game.Justify;
import cc.lib.game.Utils;
import cc.lib.swing.AWTFrame;
import cc.lib.swing.AWTKeyboardAnimationApplet;
import cc.lib.utils.FileUtils;
import marcos.games.hexes.core.Board;
import marcos.games.hexes.core.Hexes;
import marcos.games.hexes.core.Piece;
import marcos.games.hexes.core.Player;
import marcos.games.hexes.core.Shape;

public class HexesApplet extends AWTKeyboardAnimationApplet {

	static HexesApplet instance = null;

	public static void main(String [] args) {
        Utils.setDebugEnabled();
        Utils.setRandomSeed(0);
        AWTFrame frame = new AWTFrame("Hexes Debug Mode");
        AWTKeyboardAnimationApplet app = new HexesApplet(frame);
        frame.add(app);
        frame.centerToScreen(640, 480);
        app.init();
        app.start();
        app.setMillisecondsPerFrame(200);
    }   

	final Hexes game = new Hexes() {

		@Override
		protected void onGameOver(final int winner) {
		    stopGame();
			new Thread() {
				public void run() {
        			int n = JOptionPane.showConfirmDialog(
        				    frame,
        				    "Player " + winner + " wins\nPlay Again?",
        				    "Game Over",
        				    JOptionPane.YES_NO_OPTION);
        			switch (n) {
        				case JOptionPane.YES_OPTION:
        					new Thread(new Runnable() {
        						public void run() {
        							showMainMenu();
        						}
        					}).start();
        					break;
        				case JOptionPane.CLOSED_OPTION:
        				case JOptionPane.NO_OPTION:
        					System.exit(0);
        					break;
        			}
				}
			}.start();
		}

		@Override
		protected void onPiecePlayed(int pIndex, int pts) {
			Piece p = getBoard().getPiece(pIndex);
			addMessage("Player %s played a %s for %d points", p.getPlayer(), p.getType(), pts);
		}
		
	};
	final AWTFrame frame;
	final File restoreFile;
	int highlightedPiece = -1;
	final LinkedList<String> messages = new LinkedList<String>();
	
	void addMessage(String msg, Object ... args) {
		synchronized (messages) {
			messages.add(String.format(msg, args));
			while (messages.size() > 3)
				messages.removeLast();
		}
	}
	
	public HexesApplet(AWTFrame frame) {
		instance = this;
		this.frame = frame;
		File settings = FileUtils.getOrCreateSettingsDirectory(getClass());
		frame.loadFromFile(new File(settings, "hexes.properties"));
		restoreFile = new File(settings, "hexes.txt");
		if (restoreFile.exists()) {
        	if (game.tryLoadFromFile(restoreFile)) {
        	    startGame();
        	    return;
            }
		}
        showHelpPopup();
	}

	private void showHelpPopup() {
		new Thread(() -> {
            String welcome = "Welcome to Hexes!\n\n"
                    + "The Object of the game is to gain the most points by placing triangular pieces to form hexagons, triangles and diamonds.\n"
                    + "";
            JOptionPane.showConfirmDialog(frame, welcome, "Welcome!", JOptionPane.INFORMATION_MESSAGE);
            showChooseColorMenu();
        }).start();
	}

	private void showChooseColorMenu() {
		new Thread() {
			public void run() {
        		int n = JOptionPane.showOptionDialog(frame,
        			    "Choose Your Color",
        			    "Choose",
        			    JOptionPane.YES_NO_OPTION,
        			    JOptionPane.PLAIN_MESSAGE,
        			    null,
        			    new String[] { "RED", "BLUE" },
        			    null);
        		if (n == 0) {
        			game.initPlayers(new SwingPlayer(), new Player());
        		} else if (n == 1) {
        			game.initPlayers(new Player(), new SwingPlayer());
        		}
        		game.newGame();
        		startGame();
			}
		}.start();
	}

	private void showMainMenu() {
		new Thread() {
			public void run() {
        		int n = JOptionPane.showOptionDialog(frame,
        			    "Choose Game",
        			    "Choose",
        			    JOptionPane.YES_NO_OPTION,
        			    JOptionPane.PLAIN_MESSAGE,
        			    null,
        			    new String[] { "NEW", "RESTORE", "CUSTOM", "AUTO" },
        			    null);
        		switch (n) {
        			case 0:
        				showChooseColorMenu(); 
        				break;
        				
        			case 1:
        				try {
        					Hexes g = new Hexes();
        					g.loadFromFile(restoreFile);
        					game.copyFrom(g);
        					startGame();
        				} catch (Exception e) {
        					e.printStackTrace();
        					JOptionPane.showMessageDialog(frame, "Error loading restore file: " + e.getMessage());
        					restoreFile.delete();
        				}
        				break;
        				
        			case 2: // TODO: CUSTOM
        				break;
        				
        			case 3: // AUTO (computer vs. computer)
        				game.initPlayers(new Player(), new Player());
        				game.newGame();
        				startGame();
        				break;
        				

        		}
			}
		}.start();
	}
	

	void stopGame() {
	    running = false;
    }

	boolean running = false;
	void startGame() {
		if (!running) {
			grabFocus();
			synchronized (messages) {
				messages.clear();
			}
			running = true;
			new Thread() {
				public void run() {
					try {
						Utils.println("Thread running");
						while (running && !game.isGameOver()) {
							synchronized (game) {
								game.runGame();
								game.saveToFile(restoreFile);
							}
						}
					} catch (Exception e) {
						e.printStackTrace();
					}
					Utils.println("Thread exiting running=" + running + " gameover=" + game.isGameOver());
					running = false;
				}
			}.start();
		}
	}
	
	@Override
	protected void doInitialization() {
	}

	@Override
	protected void drawFrame(AGraphics g) {
		g.clearScreen(GColor.LIGHT_GRAY);
		Board b = game.getBoard();
		b.setHighlighted(getMouseX(), getMouseY());
		highlightedPiece = b.draw(g);
		g.ortho();
		g.setColor(GColor.RED);
		if (game.getNumPlayers() > 1) {
    		String redTxt = "Points     " + game.getPlayer(1).getScore() + "\n"
    				      + "Hexagons   " + game.getPlayer(1).getShapeCount(Shape.HEXAGON) + "\n"
    				      + "Triangles  " + game.getPlayer(1).getShapeCount(Shape.TRIANGLE) + "\n"
    				      + "Diamonds   " + game.getPlayer(1).getShapeCount(Shape.DIAMOND);
    		
    		g.drawJustifiedString(10, 10, Justify.LEFT, redTxt);
    		g.setColor(GColor.BLUE);
    		String blueTxt = "Points     " + game.getPlayer(2).getScore() + "\n"
    				       + "Hexagons   " + game.getPlayer(2).getShapeCount(Shape.HEXAGON) + "\n"
    				       + "Triangles  " + game.getPlayer(2).getShapeCount(Shape.TRIANGLE) + "\n"
    				       + "Diamonds   " + game.getPlayer(2).getShapeCount(Shape.DIAMOND);
    		
    		g.drawJustifiedString(g.getViewportWidth()-10, 10, Justify.RIGHT, blueTxt);
		}
		// draw messages
		int tx = g.getViewportWidth()/2;
		int ty = 10;
		g.setColor(GColor.BLACK);
		synchronized (messages) {
    		for (String msg : messages) {
    			g.drawJustifiedString(tx, ty, Justify.CENTER, msg);
    			ty += g.getTextHeight();
    		}
		}

	}
	
	@Override
	protected void onDimensionsChanged(AGraphics g, int width, int height) {
		
	}

}

File: Hexes/src/marcos/games/hexes/swing/HexesApplet.kt
package marcos.games.hexes.swing

import cc.lib.game.AGraphics
import cc.lib.game.GColor
import cc.lib.game.Justify
import cc.lib.game.Utils
import cc.lib.swing.AWTFrame
import cc.lib.swing.AWTKeyboardAnimationApplet
import cc.lib.utils.FileUtils
import marcos.games.hexes.core.Hexes
import marcos.games.hexes.core.Player
import marcos.games.hexes.core.Shape
import java.io.File
import java.util.*
import javax.swing.JOptionPane

class HexesApplet(frame: AWTFrame) : AWTKeyboardAnimationApplet() {
	val game: Hexes = object : Hexes() {
		override fun onGameOver(winner: Int) {
			stopGame()
			object : Thread() {
				override fun run() {
					val n = JOptionPane.showConfirmDialog(
						frame,
						"Player $winner wins\nPlay Again?",
						"Game Over",
						JOptionPane.YES_NO_OPTION)
					when (n) {
						JOptionPane.YES_OPTION -> Thread(Runnable { showMainMenu() }).start()
						JOptionPane.CLOSED_OPTION, JOptionPane.NO_OPTION -> System.exit(0)
					}
				}
			}.start()
		}

		override fun onPiecePlayed(pIndex: Int, pts: Int) {
			val p = board.getPiece(pIndex)
			addMessage("Player %s played a %s for %d points", p.player, p.type, pts)
		}
	}

	@JvmField
	val frame: AWTFrame
	val restoreFile: File

	@JvmField
	var highlightedPiece = -1
	val messages = LinkedList<String>()
	fun addMessage(msg: String?, vararg args: Any?) {
		synchronized(messages) {
			messages.add(String.format(msg!!, *args))
			while (messages.size > 3) messages.removeLast()
		}
	}

	private fun showHelpPopup() {
		Thread(Runnable {
			val welcome = """
	        	Welcome to Hexes!
	        	
	        	The Object of the game is to gain the most points by placing triangular pieces to form hexagons, triangles and diamonds.
	        	
	        	""".trimIndent()
			JOptionPane.showConfirmDialog(frame, welcome, "Welcome!", JOptionPane.INFORMATION_MESSAGE)
			showChooseColorMenu()
		}).start()
	}

	private fun showChooseColorMenu() {
		object : Thread() {
			override fun run() {
				val n = JOptionPane.showOptionDialog(frame,
					"Choose Your Color",
					"Choose",
					JOptionPane.YES_NO_OPTION,
					JOptionPane.PLAIN_MESSAGE,
					null, arrayOf("RED", "BLUE"),
					null)
				if (n == 0) {
					game.initPlayers(SwingPlayer(), Player())
				} else if (n == 1) {
					game.initPlayers(Player(), SwingPlayer())
				}
				game.newGame()
				startGame()
			}
		}.start()
	}

	private fun showMainMenu() {
		object : Thread() {
			override fun run() {
				val n = JOptionPane.showOptionDialog(frame,
					"Choose Game",
					"Choose",
					JOptionPane.YES_NO_OPTION,
					JOptionPane.PLAIN_MESSAGE,
					null, arrayOf("NEW", "RESTORE", "CUSTOM", "AUTO"),
					null)
				when (n) {
					0 -> showChooseColorMenu()
					1 -> try {
						val g = Hexes()
						g.loadFromFile(restoreFile)
						game.copyFrom(g)
						startGame()
					} catch (e: Exception) {
						e.printStackTrace()
						JOptionPane.showMessageDialog(frame, "Error loading restore file: " + e.message)
						restoreFile.delete()
					}
					2 -> {}
					3 -> {
						game.initPlayers(Player(), Player())
						game.newGame()
						startGame()
					}
				}
			}
		}.start()
	}

	fun stopGame() {
		running = false
	}

	var running = false

	init {
		instance = this
		this.frame = frame
		val settings = FileUtils.getOrCreateSettingsDirectory(javaClass)
		frame.loadFromFile(File(settings, "hexes.properties"))
		restoreFile = File(settings, "hexes.txt")
		if (restoreFile.exists() && game.tryLoadFromFile(restoreFile)) {
			startGame()
		} else {
			showHelpPopup()
		}
	}

	fun startGame() {
		if (!running) {
			grabFocus()
			synchronized(messages) { messages.clear() }
			running = true
			object : Thread() {
				override fun run() {
					try {
						Utils.println("Thread running")
						while (running && !game.isGameOver) {
							synchronized(game) {
								game.runGame()
								game.saveToFile(restoreFile)
							}
						}
					} catch (e: Exception) {
						e.printStackTrace()
					}
					Utils.println("Thread exiting running=" + running + " gameover=" + game.isGameOver)
					running = false
				}
			}.start()
		}
	}

	override fun doInitialization() {}
	override fun drawFrame(g: AGraphics) {
		g.clearScreen(GColor.LIGHT_GRAY)
		val b = game.board
		b.setHighlighted(mouseX, mouseY)
		highlightedPiece = b.draw(g)
		g.ortho()
		g.color = GColor.RED
		if (game.numPlayers > 1) {
			val redTxt = """
	        	Points     ${game.getPlayer(1).score}
	        	Hexagons   ${game.getPlayer(1).getShapeCount(Shape.HEXAGON)}
	        	Triangles  ${game.getPlayer(1).getShapeCount(Shape.TRIANGLE)}
	        	Diamonds   ${game.getPlayer(1).getShapeCount(Shape.DIAMOND)}
	        	""".trimIndent()
			g.drawJustifiedString(10f, 10f, Justify.LEFT, redTxt)
			g.color = GColor.BLUE
			val blueTxt = """
	        	Points     ${game.getPlayer(2).score}
	        	Hexagons   ${game.getPlayer(2).getShapeCount(Shape.HEXAGON)}
	        	Triangles  ${game.getPlayer(2).getShapeCount(Shape.TRIANGLE)}
	        	Diamonds   ${game.getPlayer(2).getShapeCount(Shape.DIAMOND)}
	        	""".trimIndent()
			g.drawJustifiedString((g.viewportWidth - 10).toFloat(), 10f, Justify.RIGHT, blueTxt)
		}
		// draw messages
		val tx = g.viewportWidth / 2
		var ty = 10
		g.color = GColor.BLACK
		synchronized(messages) {
			for (msg in messages) {
				g.drawJustifiedString(tx.toFloat(), ty.toFloat(), Justify.CENTER, msg)
				ty += g.textHeight.toInt()
			}
		}
	}

	override fun onDimensionsChanged(g: AGraphics, width: Int, height: Int) {}

	companion object {
		@JvmField
		var instance: HexesApplet? = null

		@JvmStatic
		fun main(args: Array<String>) {
			Utils.setDebugEnabled()
			Utils.setRandomSeed(0)
			val frame = AWTFrame("Hexes Debug Mode")
			val app: AWTKeyboardAnimationApplet = HexesApplet(frame)
			frame.add(app)
			frame.centerToScreen(640, 480)
			app.init()
			app.start()
			app.setMillisecondsPerFrame(200)
		}
	}
}